### ISSUE TITLE
Noisy warnings from broadcasting in boolean operations

### ISSUE THREAD

**Issue by Parskatt** [pytorch/pytorch#110941]
### 🐛 Describe the bug

The following code:
```python
import torch

I,J,K = 5,3,7

x1 = torch.zeros(I,J,1).bool()
x2 = torch.zeros(I,1,K).bool()
torch.logical_and(x1, x2)
```
raises 
```bash
UserWarning: An output with one or more elements was resized since it had shape [5, 3, 1], which does not match the required output shape [5, 3, 7]. This behavior is deprecated, and in a future PyTorch release outputs will not be resized unless they have zero elements. You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0). (Triggered internally at ../aten/src/ATen/native/Resize.cpp:28.)
  torch.logical_and(x1, x2)
```

However the equivalent

```python
import torch

I,J,K = 5,3,7

x1 = torch.zeros(I,J,1).bool()
x2 = torch.zeros(I,1,K).bool()
x1*x2
```

Does not raise any warnings. This seems inconsistent.

There seems to have been some reason (bugfix) for raising this warning, see #107431 .

### Versions

```bash
[pip3] numpy==1.26.0
[pip3] torch==2.1.0
[pip3] torch-tb-profiler==0.4.1
[pip3] triton==2.1.0
```

cc @ezyang @gchanan @zou3519 @kadeng

**Comment by Parskatt** [comment]
It's also unclear why "[5, 3, 1], which does not match the required output shape [5, 3, 7]" would even be considered a resize? It seems like standard broadcasting to me.

**Comment by cpuhrsch** [comment]
I can reproduce this and agree it's confusing. Marking this as high priority on how to resolve it and which label to associate this with.

**Comment by zou3519** [comment]
This is a regression from 2.0.0

**Comment by atalman** [comment]
@zou3519 could you please provide a fix for this issue

**Comment by huydhn** [comment]
AI: Cherry pick https://github.com/pytorch/pytorch/pull/112170

**Comment by atalman** [comment]
closing this one since resolved and cherry picked
